protocol/src/protocol.semanticHighlighting.proposed.md

```diff
@@ -0,0 +1,134 @@
+	 * feature. Otherwise, clients should reuse this "lookup table" when receiving semantic highlighting notifications from
+	 * the server.
+	 */
+	scopes?: string[][];
+}
+```
+
+##### SemanticHighlighting Notification
+
+The `textDocument/semanticHighlighting` notification is pushed from the server to the client to inform the client about additional semantic highlighting information that has to be applied on the text document. It is the server's responsibility to decide which lines are included in the highlighting information. In other words, the server is capable of sending only a delta information. For instance, after opening the text document (`DidOpenTextDocumentNotification`) the server sends the semantic highlighting information for the entire document, but if the server receives a `DidChangeTextDocumentNotification`, it pushes the information only about the affected lines in the document.
+
+The server never sends delta notifications, if no new semantic highlighting ranges were introduced but the existing onces have been shifted. For instance, when inserting a new line to the very beginning of the text document. The server receives the `DidOpenTextDocumentNotification`, updates its internal state, so that the client and the server shares the same understanding about the highlighted positions but the server does not send any notifications to the client. In such cases, it is the client's responsibility to track this event and shift all existing markers.
+
+The server can send a `SemanticHighlightingInformation` to the client without defining the `tokens` string. This means, the client must discard all semantic highlighting information in the line. For instance when commenting out a line.
+
+_Notification_:
+
+* method: 'workspace/semanticHighlighting'
+* params: `SemanticHighlightingParams` defined as follows:
+
+```ts
+/**
+ * Parameters for the semantic highlighting (server-side) push notification.
+ */
+export interface SemanticHighlightingParams {
+
+	/**
+	 * The text document that has to be decorated with the semantic highlighting information.
+	 */
+	textDocument: VersionedTextDocumentIdentifier;
+
+	/**
+	 * An array of semantic highlighting information.
+	 */
+	lines: SemanticHighlightingInformation[];
+
+}
+
+/**
+ * Represents a semantic highlighting information that has to be applied on a specific line of the text document.
+ */
+export interface SemanticHighlightingInformation {
+
+	/**
+	 * The zero-based line position in the text document.
+	 */
+	line: number;
+
+	/**
+	 * A base64 encoded string representing every single highlighted characters with its start position, length and the "lookup table" index of
+	 * of the semantic highlighting [TextMate scopes](https://manual.macromates.com/en/language_grammars).
+	 * If the `tokens` is empty or not defined, then no highlighted positions are available for the line.
+	 */
+	tokens?: string;
+
+}
+```
+
+_Tokens_:
```

Comment:
could this section please make the binary encoding of the base64 encoded data more explicit?

Basically the following things are missing in my opinion:
- the encoded string represents an array of 64-bit sized tokens
- size of the token string is thus size of the decoded string divided by 8
- what endianness is used for the unsigned integers in the token? system default? can we make this explicit and enforce little endian, assuming one can in principle run a lsp server on a big endian machine and connect to it from a little endian client or vice versa?

Reply:
regarding endianness, it seems to be defined as big endian, cf. clangd source:

```
     // Writes the token to LineByteTokens in the byte format specified by the
      // LSP proposal. Described below.
      // |<---- 4 bytes ---->|<-- 2 bytes -->|<--- 2 bytes -->|
      // |    character      |  length       |    index       |

      write32be(Token.R.start.character, OS);
      write16be(Token.R.end.character - Token.R.start.character, OS);
      write16be(static_cast<int>(Token.Kind), OS);
```